### PR TITLE
Adding localized Mur\SA-Mur absorbers for waveguide ports and custom waveguide modes

### PR DESCRIPTION
**Pull request by gadiLhv**
There are two additions in this pull requests:

1. Support for Mur B.C. locally (not necessary on the edge) as a primitive
Same use cases as specified in the CSXCAD PR
- Horn antennaspattern calculation
- Coaxial to PCB transitions.
Two types or Mur B.Cs were implemented:
- Regular first-order Mur
- Super-absorbing Mur B.C (see reference [1]). I'm pretty sure at this point that the functionality is at least similar to surface-impedance boundary conditions. They did increase convergence rate in some cases.

2. Externally calculated modes are also integrated for V\I calculations. S-Parameters verified, as well.
image

All original functionality is maintained. All functions can be used as before, as well.

[1] Betz, Vaughn Timothy, and R. Mittra. "Absorbing boundary conditions for the finite-difference time-domain analysis of guided-wave structures." Coordinated Science Laboratory Report no. UILU-ENG-93-2243 (1993).‏

### CODE REVIEW

**Comment by gadiLhv** [comment]
Added two example scripts to openEMS\Python\Tutorials

**Comment by gadiLhv** [comment]
Checked that this commit\push compiles.

Ready for instructions

**Comment by thliebig** [comment]
A few initial comments and things you need to change:
* It's obvious that you took the "normal" Mur boundary condition extension as basis, make sure to fully cleanup then, e.g. comments, names, dates and such
* The new ArrayLib infrastructure has been introduced to improve array handling since you first started . Please make use of this new array approach by once again looking how it is done in the Mur-BC extension.The current approach with a '#define" is not how I usually do it (even before the ArrayLib) and running you code also creates a "double free" and subsequent segmentation fault, showing that array cleanup is not done correctly too.
* Cleanup the large commented sections
* I recommend to move the "python test code" (MSL_With_Local_Absorbers.py) into a different folder, maybe under Tests or Examples (would need to be created). Once this is all done we should create a proper Tutorial, but this is not it.
* Maybe a (rect) waveguide would be a good showcase for this ABC?

**Comment by gadiLhv** [comment]
Thank you for reviewing this so quickliy!

> It's obvious that you took the "normal" Mur boundary condition extension as basis, make sure to fully cleanup then, e.g. comments, names, dates and such

Sure did! The lumped elements PR was based on Lorentz material, BTW. Will clean it up soon enough, as requested.

> The new ArrayLib infrastructure has been introduced to improve array handling since you first started .

That's what I get for moving slowly :). I'll look into it.

> Maybe a (rect) waveguide would be a good showcase for this ABC?

No problem, although one could claim it can be done with placing the waveguide on the boundaries... I'll figure something out. I wanted to implement a coax->waveguide transition, but the CAD work drove me nuts.

**Comment by gadiLhv** [comment]
After starting to get acquainted with the new ArrayLib, I have a question:
Unlike the original Mur BC extension, this one consists of (probably) much smaller areas to perform the operation on. What is the best approach out of these two?

1. Somehow concatenate all of the lines into one vector and input them into a single array?
2. An array of ArrayIJ, one for each boundary condition?

I don't know if this question is for @thliebig or @biergaizi, who wrote this.

Cheers

**Comment by thliebig** [comment]
I'm not sure I understand the question fully, but yes, for the 2D array I guess you should use an ArrayIJ instead. This will create a single/linear array, but can be efficiently accessed like a 2D array

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	m_numCells = 0;
+
+	for (unsigned int dirIdx = 0 ; dirIdx < 3 ; m_dir[dirIdx++] = 0);
+	for (unsigned int dirIdx = 0 ; dirIdx < 3 ; m_sheetX0[dirIdx++] = 0);
+	for (unsigned int dirIdx = 0 ; dirIdx < 3 ; m_sheetX1[dirIdx++] = 0);
+
+	m_normalSignPositive = true;
+
+	// Storage for the normal direction of each sheet
+	m_normDir = 1;
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:57]: Why is this initialized to 1? Why not to -1 to make clear it is invalid at this point?

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+	unsigned int sheetCheck = 0;
+	for (unsigned int dimIdx = 0 ; dimIdx < 3 ; dimIdx++)
+		sheetCheck += (uiStop[dimIdx] - uiStart[dimIdx] + 1) == 1;
+
+	if (sheetCheck != 1)
+	{
+		cerr 	<< "Operator_Ext_Absorbing_BC::SetInitParams(): Warning: Absorbing sheet is not a sheet! Skipping. "
+				<< " ID: " << prim->GetID() << " @ Property: " << abc_prop->GetName() << endl;
+		return false;
+	}
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:112]: SnapBox2Mesh already returns the dimension, you would only need to check that for 2 and be good?

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+	// Get box start and stop positions
+
+
+	// snap to the native coordinate system
+	int Snap_Dimension =
+		m_Op->SnapBox2Mesh(
+				cSheet->GetStartCoord()->GetCoords(m_Op->m_MeshType),	// Start Coord
+				cSheet->GetStopCoord()->GetCoords(m_Op->m_MeshType),	// Stop Coord
+				uiStart,	// Start Index
+				uiStop,		// Stop Index
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:88]: Why use two extra variables? Why not directly m_SheetX0/1 ?

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+				false,		// Dual (doublet) Grid?
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:89]: Whats doublet grid? This enables snapping to the secondary or dual or H-mesh

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+		sheetCheck += (uiStop[dimIdx] - uiStart[dimIdx] + 1) == 1;
+
+	if (sheetCheck != 1)
+	{
+		cerr 	<< "Operator_Ext_Absorbing_BC::SetInitParams(): Warning: Absorbing sheet is not a sheet! Skipping. "
+				<< " ID: " << prim->GetID() << " @ Property: " << abc_prop->GetName() << endl;
+		return false;
+	}
+
+	double phaseVelocity = abc_prop->GetPhaseVelocity();
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:114]: Again, why use a new variable and not m_phaseVelocity ?

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+	// Copy all of the relevant data, so BuildExtension ca0n work
+	m_ABCtype = (ABCtype)(abc_prop->GetAbsorbingBoundaryType());
+	m_normalSignPositive = abc_prop->GetNormalSignPositive();
+	m_phaseVelocity = phaseVelocity;
+
+	for (int cDim = 0 ; cDim < 3 ; cDim++)
+	{
+		m_sheetX0[cDim] = uiStart[cDim];
+		m_sheetX1[cDim] = uiStop[cDim];
+	}
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:133]: unneeded, see above

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+					totCells = 1;
+
+	// Initialize normDir with an unrealistic value
+	m_normDir = 0;
+	for (unsigned int dimIdx = 0 ; dimIdx < 3 ; dimIdx++)
+	{
+		Ncells[dimIdx] = m_sheetX1[dimIdx] - m_sheetX0[dimIdx] + 1;
+		totCells *= Ncells[dimIdx];
+		m_normDir = (Ncells[dimIdx] == 1) ? (dimIdx + 1) : m_normDir;
+	}
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:160]: We talked about this, the normal dir must be 0 for x, 1 for y and 2 for z to be consistent with everywhere else in openEMS! not 1 for x and so on!

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+	m_numCells = totCells;
+
+	// Update direction array
+	for (unsigned int dimIdx = 0 ; dimIdx < 3 ; dimIdx++)
+		m_dir[dimIdx] = (m_normDir + dimIdx - 1) % 3;
+
+	// Add direction sign
+	m_normDir *= (m_normalSignPositive ? 1 : -1);
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:169]: Don't do this, keep the norm dir as 0/1/2 (see above) and keep the normal sign. Why smash them into the one variable. This is not how openEMS does this...

File: FDTD/extensions/operator_ext_absorbing_bc.h [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+protected:
+
+	Operator_Ext_Absorbing_BC(Operator* op, Operator_Ext_Absorbing_BC* op_ext);
+	void Initialize();
+
+	//unsigned int			m_numPrims;		// Number of primitives that will apply absorbing B.C.
+	unsigned int			m_numCells;		// Number of cells in each primitive
+
+	// Storage for the modulated direcions. m_dir[primIdx][0] is the normal, the rest are the tangential components.
+	unsigned int	m_dir[3];
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.h:83]: Unclear, what is m_dir supposed to be?

File: FDTD/extensions/operator_ext_absorbing_bc.cpp [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+		Ncells[dimIdx] = m_sheetX1[dimIdx] - m_sheetX0[dimIdx] + 1;
+		totCells *= Ncells[dimIdx];
+		m_normDir = (Ncells[dimIdx] == 1) ? (dimIdx + 1) : m_normDir;
+	}
+
+	m_numCells = totCells;
+
+	// Update direction array
+	for (unsigned int dimIdx = 0 ; dimIdx < 3 ; dimIdx++)
+		m_dir[dimIdx] = (m_normDir + dimIdx - 1) % 3;
```
Comment (thliebig) [inline FDTD/extensions/operator_ext_absorbing_bc.cpp:166]: This is not how it is supposed to be done? Use variables like normP (norm+1) or nyP in short and nyPP (norm+2) instead. This way m_dir is not necessary at all and this is all more consistent...

**Review by thliebig — CHANGES_REQUESTED** [review]
There are quite a lot of issues still. Mostly concerning the openEMS internal conventions about the directions. I have not reviewed everything, e.g. the engine extensions part.
Please check them too.

File: FDTD/extensions/operator_ext_absorbing_bc.cpp (hunk 2 above)
Comment (gadiLhv) [reply FDTD/extensions/operator_ext_absorbing_bc.cpp:112]: Sorry, I didn't really understand this one. 
I got the dimensions from SnapBox2Mesh originally. What I'm checking is that this isn't a volume. 
You mean Only check the first two dimensions?

File: FDTD/extensions/operator_ext_absorbing_bc.cpp (hunk 4 above)
Comment (gadiLhv) [reply FDTD/extensions/operator_ext_absorbing_bc.cpp:89]: Yeah well, you tell Prof. Shavit that (the guy who taught me this course). That's how he called it back then.
I'll remove the comment, due to the fact that the course was horrible.

File: FDTD/extensions/operator_ext_absorbing_bc.h (hunk 9 above)
Comment (gadiLhv) [reply FDTD/extensions/operator_ext_absorbing_bc.h:83]: I'm not sure it's a good idea to make this a memeber. Need to see references from other extensions and I'll decide what to do.

**Comment by gadiLhv** [comment]
I think I took care of all of them @thliebig 
I had a mess while pushing this, hope I didn't break anything else.

**Comment by biergaizi** [comment]
I suggest manually cleaning up the commit history, otherwise these "rebase errors" will make it difficult to track project history in the future. I usually do this:

1. Fetch and checkout a fresh upstream main/master branch

    git add remote upstream https://github.com/thliebig/openEMS.git
    git fetch upstream/master

2. Create a fresh fork of the main branch.

    git checkout upstream/master
    git branch feature-clean
    git checkout feature-clean
    
3. Use `git cherry-pick` to manually grab the commits one by one from the "dirty" branch.

    git log feature-dirty
    # read the commit log, copy the hashes of the earliest commit
    git cherry-pick $hash

4. See if this patch is clean, or if it contains errors. If it contains errors, immediately fix it and commit it.

    git commit source.c
    
5. Cherry-pick the next patch, and repeat.

6. Now the commit log looks like the following:

    original commit 1
    fixup 1
    original commit 2
    fixup 2
    ....

7. Merge this fixups into the parent commit by using the "fixup" mode of git rebase:

    git rebase -i first_sha~1
    
8. Change the keyword of the fixup commits as "fixup".

9. Save and exit.

Now the commit history has been rewritten and cleaned up.

**Comment by gadiLhv** [comment]
I'll try, @biergaizi 
I did a mistake and forgot to fetch this branch. Somehow over-written the entire thing, and did another sin by force pushing instead of reverting...

If this is a problem, maybe it's better that I withdraw this PR and start a new one? 
Let me know, @biergaizi or @thliebig

**Comment by biergaizi** [comment]
I believe you can change your source branch even after sending a PR, can't you. Just prepare a clean branch and edit your PR when you're ready.

**Comment by gadiLhv** [comment]
Ok, @thliebig and @biergaizi 
I squashed everyhing after rebasing (yet again) against the master from this repo.
The next PR (waveguide) is already in it's own branch, so everything should be cleaner in the future. I am learning this "on the fly" (because "on job" is incorrect, here...)

**Comment by gadiLhv** [comment]
Hey @biergaizi 
I didn't find a way to switch this PR to a new branch, sorry. It's actually possible to change the base branch, but that doesn't really help.
image

**Comment by thliebig** [comment]
I'm not sure what the issue is you two are discussing, but the PR-tree here must not be in perfect shape. I guess once I accept the PR I can just squash merge everything and make it neat and tidy?

**Comment by gadiLhv** [comment]
Yes, and I promise to be a good pseudo-programmer and keep everything in appropriate branches from now on.

**Comment by biergaizi** [comment]
Keep up the good work. As a programmer, I'm also trying to be a better pseudo-engineer and pseudo-physicist as well.

**Comment by thliebig** [comment]
While you are editing, maybe you also want to change some of the copyright/credit lines in one of the *.py and one of the *.h files :smirk:

And you do not need the metal boxes in the rect waveguide test? You have set the normal boundary conditions to PEC anywhere anyhow?
